README.md
# **MKANET V2 - The ultimate gray color theme**

I decided to create my own gray color theme from scratch to avoid headaches I got from traditional MKANET and light-colored themes.  I spent a lot of time choosing precise colors for syntax coloring and the Sidebar colors and customizing UI components.

Below, are the bonus features included with this theme:
<br/>
<br/>


## **Vibrancy sidebar:**
 
![Theme Screenshot](https://github.com/mkanet/MKANET-Theme-V2/raw/master/./screenshots/screenshot1.png)
<br/>
<br/>


## **Custom code folding icons:** 

![Theme Screenshot](https://github.com/mkanet/MKANET-Theme-V2/raw/master/./screenshots/screenshot2.gif)
<br/>
<br/>


## **Custom tab close icons:**

![Theme Screenshot](https://github.com/mkanet/MKANET-Theme-V2/raw/master/./screenshots/screenshot3.gif)
<br/>
<br/>


## **Supported Operating Systems:**

Windows 10 ✔

MacOS ✔
<br/>
<br/>


## **Available Vibrancy themes:**

* **Dark** &nbsp;*(default)*
* **Darker**

|        Dark      |    Darker   |
|:------------------------:|:-----------------------:|
| ![](https://github.com/mkanet/MKANET-Theme-V2/raw/master/./screenshots/Dark.png) | ![](https://github.com/mkanet/MKANET-Theme-V2/raw/master/./screenshots/Darker.png) | 


<br/>

**Designer:&nbsp;&nbsp;** Michael K. Avanessian

**Special Thanks To:&nbsp;&nbsp;** *"illixion"* from Github:&nbsp;[ILLIXION/vscode-vibrancy-continued](https://github.com/ILLIXION/vscode-vibrancy-continued) *(for fixing original Vibrancy Extension "Mouse lag" and "Flashing" bux fixes in VSCode >1.56.2)* ⭐

**Special Thanks To:&nbsp;&nbsp;** *"eyhn"* from Github:&nbsp;[EYHN/vscode-vibrancy](https://github.com/EYHN/vscode-vibrancy) *(for original Vibrancy and custom CSS code)* ⭐

CHANGELOG.md
# Changes

* 0.3.6 - Minor color adjustments to Dark & Darker themes
* 0.3.5 - Fix: compatibility issues with vscode >1.56.2
* 0.3.4 - Fix: Avoid getting "Your Code installation appears to be corrupt. Please reinstall." message.
* 0.3.3 - Fix: compatibility issues with vscode 1.56.0.  Changed base theme to from 'vs-dark' to 'vs'
* 0.3.2 - Changed terminal ansi text colors to original retro colors
* 0.3.1 - Fix: compatibility issue with vscode 1.53.0
* 0.3.0 - Fine tuned text colors
* 0.2.9 - Fixed [Vibrancy blur mouse dragging lag issue](https://github.com/EYHN/vscode-vibrancy/discussions/80)
* 0.2.8 - Fixed sidebar "dirty circle" + "close" icons; as well as pencil icon
* 0.2.7 - Fixed tab icons for compatibilty with VSCode ver. 1.51.1 and higher
* 0.2.6 - Fixed sidebar transparency
* 0.2.5 - Fixed code folding/collapsing icons.  Fixed settings pencil icon
* 0.2.4 - Adjusted "dirty circle" position
* 0.2.3 - Adjusted "tab-close" position
* 0.2.2 - Resized "tab-close" icons to match "dirty circle"
* 0.2.1 - Changed spacing for tab close icons
* 0.2.0 - More minor adjustments
* 0.1.9 - Minor custom icon size adjustments
* 0.1.8 - Adjusted site for code folding icon
* 0.1.7 - Changes to custom icons for compatibility with VSCode ver. 1.39.1 and higher
* 0.1.6 - Added settings pencil icon.  Updated a couple of syntax highlighting colors
* 0.1.5 - Updated to latest Vibrancy release.  Fixed code folding icon bug.
* 0.1.4 - Fixed bug in previous update
* 0.1.3 - Added "Lighter" vibrancy theme
* 0.1.2 - Added Panel vibrancy
* 0.1.1 - Reverted back to previous Function parameter color
* 0.1.1 - Reverted back to previous Function parameter color
* 0.1.0 - Modified Function parameter syntax highlighting color
* 0.0.9 - Recreated tab close icon with a better one
* 0.0.8 - Replaced tab close icon
* 0.0.7 - Added custom tab close icons and syntax highlighting enhancements
* 0.0.6 - First public release


runtime/index.js
const electron = require('electron');
/**
 * @type {(window) => Record<'interval' | 'overwrite', {install: () => void, uninstall: () => void>}
 */
const transparencyMethods = require('./methods');

/**
 * @type {{
 *  os: string,
 *  config: {
 *    type:  "auto" | "acrylic" | "appearance-based" | "light" | "dark" | "titlebar" | "selection" | "menu" | "popover" | "sidebar" | "medium-light" | "ultra-dark",
 *    opacity: number,
 *    theme: "Dark" | "Darker",
 *    imports: string[],
 *    refreshInterval: number,
 *    preventFlash: boolean
 *  },
 *  themeCSS: string,
 *  theme: any,
 *  imports: {
 *    css: string,
 *    js: string
 *  }
 * }}
 */
const app = global.vscode_vibrancy_plugin;
// @ts-check

const macosType = [
  'appearance-based',
  'light',
  'dark',
  'titlebar',
  'selection',
  'menu',
  'popover',
  'sidebar',
  'medium-light',
  'ultra-dark',
];

const windowsType = ['acrylic'];

/**
 * @param {string} hex
 * @returns {{ r: any; g: any; b: any; } | null}
 */
function hexToRgb(hex) {
  var result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
  return result
    ? {
      r: parseInt(result[1], 16),
      g: parseInt(result[2], 16),
      b: parseInt(result[3], 16),
    }
    : null;
}

electron.app.on('browser-window-created', (_, window) => {
  const methods = transparencyMethods(window);
  const hackMethod = app.config.preventFlash ? 'overwrite' : 'interval';
  const effects = methods[hackMethod];

  var type = app.config.type;
  if (type !== 'auto') {
    if (app.os === 'win10' && !windowsType.includes(type)) type = 'auto';
    if (app.os === 'macos' && !macosType.includes(type)) type = 'auto';
  }
  if (type === 'auto') {
    type = app.theme.type[app.os];
  }

  let opacity = app.config.opacity;
  // if opacity < 0, use the theme default opacity
  if (opacity < 0) {
    opacity = app.theme.opacity[app.os];
  }

  const backgroundRGB = hexToRgb(app.theme.background) || { r: 0, g: 0, b: 0 };

  if (app.os === 'win10') {
    const bindings = require('./vibrancy.js');
    bindings.setVibrancy(
      window.getNativeWindowHandle().readInt32LE(0),
      1,
      backgroundRGB.r,
      backgroundRGB.g,
      backgroundRGB.b,
      0
    );
    const win10refresh = require('./win10refresh.js');
    win10refresh(window, 60);

    window.webContents.once('dom-ready', () => {
      const currentURL = window.webContents.getURL();

      if (
        !(
          currentURL.includes('workbench.html') ||
          currentURL.includes('workbench-monkey-patch.html')
        )
      ) {
        return;
      }

      if (window.isMaximized()) {
        window.unmaximize();
        window.maximize();
      }
    });
  }

  window.on('closed', () => {
    effects.uninstall();
  });

  window.webContents.on('dom-ready', () => {
    const currentURL = window.webContents.getURL();

    if (
      !(
        currentURL.includes('workbench.html') ||
        currentURL.includes('workbench-monkey-patch.html')
      )
    ) {
      return;
    }

    window.setBackgroundColor('#00000000');

    effects.install();

    if (app.os === 'macos') {
      window.setVibrancy(type);

      // hack
      const width = window.getBounds().width;
      window.setBounds({
        width: width + 1,
      });
      window.setBounds({
        width,
      });
    }

    injectHTML(window);
  });
});

function injectHTML(window) {
  window.webContents.executeJavaScript(`(function(){
    const vscodeVibrancyTTP = window.trustedTypes.createPolicy("VscodeVibrancy", { createHTML (v) { return v; }});

    document.getElementById("vscode-vibrancy-style")?.remove();
    const styleElement = document.createElement("div");
    styleElement.id = "vscode-vibrancy-style";
    styleElement.innerHTML = vscodeVibrancyTTP.createHTML(${JSON.stringify(
    styleHTML()
  )});
    document.body.appendChild(styleElement);

    document.getElementById("vscode-vibrancy-script")?.remove();
    const scriptElement = document.createElement("div");
    scriptElement.id = "vscode-vibrancy-script";
    scriptElement.innerHTML = vscodeVibrancyTTP.createHTML(${JSON.stringify(
    scriptHTML()
  )});
    document.body.appendChild(scriptElement);
  })();`);
}


function scriptHTML() {
  return app.imports.js;
}

function styleHTML() {
  if (app.os === 'unknown') return '';

  var type = app.config.type;
  if (type === 'auto') {
    type = app.theme.type[app.os];
  }

  let opacity = app.config.opacity;

  if (opacity < 0) {
    opacity = app.theme.opacity[app.os];
  }

  const backgroundRGB = hexToRgb(app.theme.background) || { r: 0, g: 0, b: 0 };

  const HTML = [
    `
    <style>
      html {
        background: rgba(${backgroundRGB.r},${backgroundRGB.g},${backgroundRGB.b},${opacity}) !important;
      }
      ${app.themeCSS}
    </style>
    `,
    app.imports.css,
  ];

  return HTML.join('');
}


runtime/win10refresh.js
const { VerticalRefreshRateContext } = require('./displayconfig')
const electron = require('electron')
const process = require('process')

function sleep(duration) {
	return new Promise(resolve => setTimeout(resolve, duration));
}

function areBoundsEqual(left, right) {
	return left.height === right.height
		&& left.width === right.width
		&& left.x === right.x
		&& left.y === right.y;
}

const billion = 1000 * 1000 * 1000;

function hrtimeDeltaForFrequency(freq) {
	return BigInt(Math.ceil(billion / freq));
}

let disableJitterFix = false

// Detect if cursor is near the screen edge. Used to disable the jitter fix in 'move' event.
function isInSnapZone() {
	const point = electron.screen.getCursorScreenPoint()
	const display = electron.screen.getDisplayNearestPoint(point)

	// Check if cursor is near the left/right edge of the active display
	return (point.x > display.bounds.x - 20 && point.x < display.bounds.x + 20) || (point.x > display.bounds.x + display.bounds.width - 20 && point.x < display.bounds.x + display.bounds.width + 20);
}

/**
 * Unfortunately, we have to re-implement moving and resizing.
 * Enabling vibrancy slows down the window's event handling loop to the
 * point building a mouse event backlog. If you just handle these events
 * in the backlog without taking the time difference into consideration,
 * you end up with visible movement lag.
 * We tried pairing 'will-move' with 'move', but Electron actually sends the
 * 'move' events _before_ Windows actually commits to the operation. There's
 * likely some queuing going on that's getting backed up. This is not the case
 * with 'will-resize' and 'resize', which need to use the default behavior
 * for compatibility with soft DPI scaling.
 * The ideal rate of moving and resizing is based on the vertical sync
 * rate: if your display is only fully updating at 120 Hz, we shouldn't
 * be attempting to reset positions or sizes any faster than 120 Hz.
 * If we were doing this in a browser context, we would just use
 * requestAnimationFrame and call it a day. But we're not inside of a
 * browser context here, so we have to resort to clever hacks.
 * This VerticalRefreshRateContext maps a point in screen space to the
 * vertical sync rate of the display(s) actually handing that point.
 * It handles multiple displays with varying vertical sync rates,
 * and changes to the display configuration while this process is running.
 */
module.exports = function win10refresh(win, maximumRefreshRate) {
	if (!('__electron_acrylic_window__' in win)) {
		win.__electron_acrylic_window__ = {};
	}

	const refreshCtx = new VerticalRefreshRateContext()

	function getRefreshRateAtCursor(cursor) {
		cursor = cursor || electron.screen.getCursorScreenPoint()
		return refreshCtx.findVerticalRefreshRateForDisplayPoint(cursor.x, cursor.y)
	}

	// Ensure all movement operation is serialized, by setting up a continuous promise chain
	// All movement operation will take the form of
	//
	//     boundsPromise = boundsPromise.then(() => { /* work */ })
	//
	// So that there are no asynchronous race conditions.
	let pollingRate
	let doFollowUpQuery = false, isMoving = false, shouldMove = false
	let moveLastUpdate = BigInt(0), resizeLastUpdate = BigInt(0)
	let lastWillMoveBounds, lastWillResizeBounds,
		desiredMoveBounds
	let boundsPromise = Promise.race([
		getRefreshRateAtCursor(electron.screen.getCursorScreenPoint()).then(rate => {
			pollingRate = rate || 30
			doFollowUpQuery = true
		}),
		// Establishing the display configuration can fail; we can't
		// just block forever if that happens. Instead, establish
		// a fallback polling rate and hope for the best.
		sleep(2000).then(() => {
			pollingRate = pollingRate || 30
		})
	])

	async function doFollowUpQueryIfNecessary(cursor) {
		if (doFollowUpQuery) {
			const rate = await getRefreshRateAtCursor(cursor)
			if (rate != pollingRate) console.log(`New polling rate: ${rate}`)
			pollingRate = rate || 30
		}
	}

	function setWindowBounds(bounds) {
		if (win.isDestroyed()) return
		win.setBounds(bounds)
		desiredMoveBounds = win.getBounds()
	}

	function currentTimeBeforeNextActivityWindow(lastTime, forceFreq) {
		return process.hrtime.bigint() <
			lastTime + hrtimeDeltaForFrequency(forceFreq || pollingRate || 30)
	}

	function guardingAgainstMoveUpdate(fn) {
		if (pollingRate === undefined || !currentTimeBeforeNextActivityWindow(moveLastUpdate)) {
			moveLastUpdate = process.hrtime.bigint()
			fn()
			return true
		} else {
			return false
		}
	}

	win.on('will-move', (e, newBounds) => {
		if (win.__electron_acrylic_window__.opacityInterval) return
		// We get a _lot_ of duplicate bounds sent to us in this event.
		// This messes up our timing quite a bit.
		if (lastWillMoveBounds !== undefined && areBoundsEqual(lastWillMoveBounds, newBounds)) {
			e.preventDefault()
			return
		}
		if (lastWillMoveBounds) {
			newBounds.width = lastWillMoveBounds.width
			newBounds.height = lastWillMoveBounds.height
		}
		lastWillMoveBounds = newBounds
		// If we're asked to perform some move update and it's under
		// the refresh speed limit, we can just do it immediately.
		// This also catches moving windows with the keyboard.
		const didOptimisticMove = !isMoving && guardingAgainstMoveUpdate(() => {
			// Do nothing, the default behavior of the event is exactly what we want.
			desiredMoveBounds = undefined
		})
		if (didOptimisticMove) {
			boundsPromise = boundsPromise.then(doFollowUpQueryIfNecessary)
			return
		}
		e.preventDefault()

		// Track if the user is moving the window
		if (win.__electron_acrylic_window__.moveTimeout) clearTimeout(win.__electron_acrylic_window__.moveTimeout)
		win.__electron_acrylic_window__.moveTimeout = setTimeout(() => {
			shouldMove = false
		}, 1000 / Math.min(pollingRate, maximumRefreshRate))

		// Disable next event ('move') if cursor is near the screen edge
		disableJitterFix = isInSnapZone()

		// Start new behavior if not already
		if (!shouldMove) {
			shouldMove = true

			if (isMoving) return false
			isMoving = true

			// Get start positions
			const basisBounds = win.getBounds()
			const basisCursor = electron.screen.getCursorScreenPoint()

			// Handle polling at a slower interval than the setInterval handler
			function handleIntervalTick(moveInterval) {
				boundsPromise = boundsPromise.then(() => {
					if (!shouldMove) {
						isMoving = false
						clearInterval(moveInterval)
						return
					}

					const cursor = electron.screen.getCursorScreenPoint()
					const didIt = guardingAgainstMoveUpdate(() => {
						// Set new position
						if (lastWillResizeBounds && lastWillResizeBounds.width) setWindowBounds({
							x: Math.floor(basisBounds.x + (cursor.x - basisCursor.x)),
							y: Math.floor(basisBounds.y + (cursor.y - basisCursor.y)),
							width: Math.floor(lastWillResizeBounds.width / electron.screen.getDisplayMatching(basisBounds).scaleFactor),
							height: Math.floor(lastWillResizeBounds.height / electron.screen.getDisplayMatching(basisBounds).scaleFactor)
						})
						else setWindowBounds({
							x: Math.floor(basisBounds.x + (cursor.x - basisCursor.x)),
							y: Math.floor(basisBounds.y + (cursor.y - basisCursor.y)),
							width: Math.floor(lastWillMoveBounds.width / electron.screen.getDisplayMatching(basisBounds).scaleFactor),
							height: Math.floor(lastWillMoveBounds.height / electron.screen.getDisplayMatching(basisBounds).scaleFactor)
						})
					})
					if (didIt) {
						return doFollowUpQueryIfNecessary(cursor)
					}
				})
			}

			// Poll at 600hz while moving window
			const moveInterval = setInterval(() => handleIntervalTick(moveInterval), 1000 / 600)
		}
	})

	win.on('move', (e) => {
		if (disableJitterFix) {
			return false
		}
		if (isMoving || win.isDestroyed()) {
			e.preventDefault()
			return false
		}
		// As insane as this sounds, Electron sometimes reacts to prior
		// move events out of order. Specifically, if you have win.setBounds()
		// twice, then for some reason, when you exit the move state, the second
		// call to win.setBounds() gets reverted to the first call to win.setBounds().
		//
		// Again, it's nuts. But what we can do in this circumstance is thwack the
		// window back into place just to spite Electron. Yes, there's a shiver.
		// No, there's not much we can do about it until Electron gets their act together.
		if (desiredMoveBounds !== undefined) {
			const forceBounds = desiredMoveBounds
			desiredMoveBounds = undefined
			win.setBounds({
				x: Math.floor(forceBounds.x),
				y: Math.floor(forceBounds.y),
				width: Math.floor(forceBounds.width),
				height: Math.floor(forceBounds.height)
			})
		}
	})

	win.on('will-resize', (e, newBounds) => {
		if (lastWillResizeBounds !== undefined && areBoundsEqual(lastWillResizeBounds, newBounds)) {
			e.preventDefault()
			return
		}

		lastWillResizeBounds = newBounds

		// 60 Hz ought to be enough... for resizes.
		// Some systems have trouble going 120 Hz, so we'll just take the lower
		// of the current pollingRate and 60 Hz.
		if (pollingRate !== undefined &&
			currentTimeBeforeNextActivityWindow(resizeLastUpdate, Math.min(pollingRate, maximumRefreshRate))) {
			e.preventDefault()
			return false
		}
		// We have to count this twice: once before the resize,
		// and once after the resize. We actually don't have any
		// timing control around _when_ the resize happened, so
		// we have to be pessimistic.
		resizeLastUpdate = process.hrtime.bigint()
	})

	win.on('resize', () => {
		resizeLastUpdate = process.hrtime.bigint()
		boundsPromise = boundsPromise.then(doFollowUpQueryIfNecessary)
	})

	// Close the VerticalRefreshRateContext so Node can exit cleanly
	win.on('closed', refreshCtx.close)
}


runtime/methods/interval.js
/**
 * @type {{
 *  os: string,
 *  config: {
 *    type:  "auto" | "acrylic" | "appearance-based" | "light" | "dark" | "titlebar" | "selection" | "menu" | "popover" | "sidebar" | "medium-light" | "ultra-dark",
 *    opacity: number,
 *    theme: "Dark" | "Darker",
 *    imports: string[],
 *    refreshInterval: number,
 *    preventFlash: boolean
 *  },
 *  themeCSS: string,
 *  theme: any,
 *  imports: {
 *    css: string,
 *    js: string
 *  }
 * }}
 */
const app = global.vscode_vibrancy_plugin;
let backgroundColorTimer;
module.exports = (window) => ({
	install() {
		clearInterval(backgroundColorTimer);
		// https://github.com/microsoft/vscode/blob/9f8431f7fccf7a048531043eb6b6d24819482781/src/vs/platform/theme/electron-main/themeMainService.ts#L80
		backgroundColorTimer = setInterval(() => {
			window.setBackgroundColor('#00000000');
		}, app.config.refreshInterval);
	},
	uninstall() {
		clearInterval(backgroundColorTimer);
	}
})


src/extension.js
var vscode = require('vscode');
var fs = require('mz/fs');
var path = require('path');
var lockPath = path.join(__dirname, '../firstload.lock');

const i18nMessages = {
	'en': JSON.parse(fs.readFileSync(path.join(__dirname, 'extension.nls.json')))
};
const defaultLocale = 'en';
const locale = (vscode.env.language || defaultLocale).toLowerCase();
const localize = (info) => {
	if (locale in i18nMessages && info in i18nMessages[locale]) {
		return i18nMessages[locale][info]
	} else {
		return i18nMessages[defaultLocale][info]
	}
}

let os = 'macos';
if (/^win/.test(process.platform)) {
	if (require('os').release().split(".").map(Number)[0] === 10) {
		os = 'win10';
	} else {
		os = 'win7';
	}
}

var themeStylePaths = {
	'Dark': '../themes/Dark.css',
	'Lighter': '../themes/Lighter.css',
}

const themeConfigPaths = {
	'Dark': '../themes/Dark.json',
	'Lighter': '../themes/Lighter.json',
}

var defaultTheme = 'Dark';

function getCurrentTheme(config) {
	return config.theme in themeStylePaths ? config.theme : defaultTheme;
}

async function changeTerminalRendererType() {
	// This is a hacky way to display the restart prompt
	let v = vscode.workspace.getConfiguration().inspect("terminal.integrated.rendererType");
	if (v !== undefined) {
		if (!v.globalValue) {
			await vscode.workspace.getConfiguration().update("terminal.integrated.rendererType", "dom", vscode.ConfigurationTarget.Global);
		}
	}
}

async function promptRestart() {
	let v = vscode.workspace.getConfiguration().inspect("window.titleBarStyle");
	if (v !== undefined) {
		let value = vscode.workspace.getConfiguration().get("window.titleBarStyle");
		await vscode.workspace.getConfiguration().update("window.titleBarStyle", value === "native" ? "custom" : "native", vscode.ConfigurationTarget.Global);
		vscode.workspace.getConfiguration().update("window.titleBarStyle", v.globalValue, vscode.ConfigurationTarget.Global);
	}
}

async function checkColorTheme() {
	const currentTheme = getCurrentTheme(vscode.workspace.getConfiguration("vscode_vibrancy"));
	const themeConfig = require(path.join(__dirname, themeConfigPaths[currentTheme]));
	const target = themeConfig.colorTheme;
	const currentColorTheme = vscode.workspace.getConfiguration().get("workbench.colorTheme");
	if (target !== currentColorTheme) {
		const message = localize('messages.recommendedColorTheme').replace('%1', currentColorTheme).replace('%2', target);
		await vscode.window.showInformationMessage(message, localize('messages.changeColorThemeIde'), localize('messages.noIde'))
			.then(async (msg) => {
				if (msg === localize('messages.changeColorThemeIde')) {
					await vscode.workspace.getConfiguration().update("workbench.colorTheme", target, vscode.ConfigurationTarget.Global);
				}
			});
	}
}

function deepEqual(obj1, obj2) {

	if(obj1 === obj2) // it's just the same object. No need to compare.
			return true;

	if(isPrimitive(obj1) && isPrimitive(obj2)) // compare primitives
			return obj1 === obj2;

	if(Object.keys(obj1).length !== Object.keys(obj2).length)
			return false;

	// compare objects with same number of keys
	for(let key in obj1)
	{
			if(!(key in obj2)) return false; //other object doesn't have this prop
			if(!deepEqual(obj1[key], obj2[key])) return false;
	}

	return true;
}

function hexToRgb(hex) {
  var result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
  return result ? {
    r: parseInt(result[1], 16),
    g: parseInt(result[2], 16),
    b: parseInt(result[3], 16)
  } : null;
}

//check if value is primitive
function isPrimitive(obj)
{
	return (obj !== Object(obj));
}

function isFirstload() {
	try {
		fs.readFileSync(lockPath);
		return false
	} catch (err) {
		return true
	}
}

function lockFirstload() {
	fs.writeFileSync(lockPath, '', () => {});
}

function injectHTML(config, currentTheme, themeConfig) {
	var type = config.type;
	if (type === 'auto') {
		type = themeConfig.type[os];
	}

	let opacity = config.opacity;

	if (opacity < 0) {
		opacity = themeConfig.opacity[os]
	}

	if (os === 'win10') {
		opacity = 0;
	}

	const backgroundRGB = hexToRgb(themeConfig.background);

	const HTML = [
		`
		<style>
			html {
				background: rgba(${backgroundRGB.r},${backgroundRGB.g},${backgroundRGB.b},${opacity}) !important;
			}
		</style>
		`,
		...config.imports.map(function (x) {
			if (!x) return;
			if (typeof x === 'string') {
				x = x.replace('%theme-style%', path.join(__dirname, themeStylePaths[currentTheme]));
				x = x.replace('$theme-style$', path.join(__dirname, themeStylePaths[currentTheme]));
				x = new URL(x, 'file://').href;

				if (!x.startsWith('file://')) {
					x = 'file://' + x;
				}
				
				if (/^.*\.js$/.test(x)) return '<script src="' + x + '"></script>';
				if (/^.*\.css$/.test(x)) return '<link rel="stylesheet" href="' + x + '"/>';
			}
		})
	]

	return HTML.join('')
}

const macosType = [
	"appearance-based",
	"light", 
	"dark", 
	"titlebar", 
	"selection", 
	"menu", 
	"popover", 
	"sidebar", 
	"medium-light", 
	"ultra-dark"
];

const windowsType = [
	"dwm",
	"acrylic"
];

function injectJS(config, currentTheme, themeConfig) {
	var type = config.type;
	if (type !== 'auto') {
		if (os === 'win10' || os === 'win7' && !windowsType.includes(type)) type = 'auto';
		if (os === 'macos' && !macosType.includes(type)) type = 'auto';
	}
	if (type === 'auto') {
		type = themeConfig.type[os];
	}

	let opacity = config.opacity;

	if (opacity < 0) {
		opacity = themeConfig.opacity[os]
	}
	
	return `
	const electron = require('electron');

  electron.app.on('browser-window-created', (event, window) => {
    window.webContents.on('dom-ready', () => {
      window.setBackgroundColor('#00000000');

      ${os !== 'macos' ? 
				`require("child_process")
					.spawn(${JSON.stringify(__dirname + '\\blur-cli.exe')}, [new Uint32Array(window.getNativeWindowHandle().buffer)[0], '--type', ${JSON.stringify(type)}, '--enable', 'true', '--color', '${themeConfig.background}', '--opacity', ${JSON.stringify(opacity)}]);` :
				`window.setVibrancy(${JSON.stringify(type)});`
			}
			
      // hack
      const width = window.getBounds().width;
      window.setBounds({
          width: width + 1,
      });
      window.setBounds({
          width,
      });

      window.webContents.executeJavaScript(${JSON.stringify("document.body.innerHTML += " + JSON.stringify(injectHTML(config, currentTheme, themeConfig)))})
    });
  })
	`
}

function activate(context) {
	console.log('vscode-vibrancy is active!');

	process.on('uncaughtException', function (err) {
		if (/ENOENT|EACCES|EPERM/.test(err.code)) {
			vscode.window.showInformationMessage(localize('messages.admin'));
			return;
		}
	});

	var isWin = /^win/.test(process.platform);
	var appDir = path.dirname(require.main.filename);

	var HTMLFile = appDir + (isWin ? '\\vs\\code\\electron-browser\\workbench\\workbench.html' : '/vs/code/electron-browser/workbench/workbench.html');
	var JSFile = appDir + (isWin ? '\\main.js' : '/main.js');

	async function installJS() {
		const config = vscode.workspace.getConfiguration("vscode_vibrancy");
		const currentTheme = getCurrentTheme(config);
		const themeConfig = require(path.join(__dirname, themeConfigPaths[currentTheme]));

		const JS = await fs.readFile(JSFile, 'utf-8');
		const newJS = JS.replace(/\/\* !! VSCODE-VIBRANCY-START !! \*\/[\s\S]*?\/\* !! VSCODE-VIBRANCY-END !! \*\//, '')
			+ '\n/* !! VSCODE-VIBRANCY-START !! */\n(function(){' + injectJS(config, currentTheme, themeConfig) + '})()\n/* !! VSCODE-VIBRANCY-END !! */\n';
		await fs.writeFile(JSFile, newJS, 'utf-8');
	}

	async function uninstallJS() {
		const JS = await fs.readFile(JSFile, 'utf-8');
		const needClean = /\/\* !! VSCODE-VIBRANCY-START !! \*\/[\s\S]*?\/\* !! VSCODE-VIBRANCY-END !! \*\//.test(JS);
		if (needClean) {
			const newJS = JS
				.replace(/\/\* !! VSCODE-VIBRANCY-START !! \*\/[\s\S]*?\/\* !! VSCODE-VIBRANCY-END !! \*\//, '')
			await fs.writeFile(JSFile, newJS, 'utf-8');
		}
	}

	async function uninstallHTML() {
		const HTML = await fs.readFile(HTMLFile, 'utf-8');
		const needClean = /<!-- !! VSCODE-VIBRANCY-START !! -->[\s\S]*?<!-- !! VSCODE-VIBRANCY-END !! -->/.test(HTML);
		if (needClean) {
			const newHTML = HTML
				.replace(/<!-- !! VSCODE-VIBRANCY-START !! -->[\s\S]*?<!-- !! VSCODE-VIBRANCY-END !! -->/, '')
				.replace(/<meta.*http-equiv="Content-Security-Policy".*>/, '');
			await fs.writeFile(HTMLFile, newHTML, 'utf-8');
		}
	}

	function enabledRestart() {
		vscode.window.showInformationMessage(localize('messages.enabled'), { title: localize('messages.reloadIde') })
			.then(function (msg) {
				msg && promptRestart();
			});
	}

	function disabledRestart() {
		vscode.window.showInformationMessage(localize('messages.disabled'), { title: localize('messages.restartIde') })
			.then(function (msg) {
				msg && promptRestart();
			});
	}

	// ####  main commands ######################################################

	async function Install(autoreload) {
		try {
			await fs.stat(JSFile);
			await changeTerminalRendererType()
		} catch (error) {
			vscode.window.showInformationMessage(localize('messages.smthingwrong') + error);
			throw error;
		}

		try {
			await installJS();
		} catch (error) {
			vscode.window.showInformationMessage(localize('messages.admin'));
			throw error;
		}
	}

	async function Uninstall() {
		try {
			await fs.stat(JSFile);
			await fs.stat(HTMLFile);
		} catch(error) {
			vscode.window.showInformationMessage(localize('messages.smthingwrong') + error);
			throw error;
		}
		
		try {
			await uninstallHTML();
			await uninstallJS();
		} catch(error) {
			vscode.window.showInformationMessage(localize('messages.admin'));
			throw error;
		}
	}

	async function Update() {
		await Uninstall();
		await Install();
	}

	var installVibrancy = vscode.commands.registerCommand('extension.installVibrancy', async () => {
		await Install();
		enabledRestart();
	});
	var uninstallVibrancy = vscode.commands.registerCommand('extension.uninstallVibrancy', async () => {
		await Uninstall()
		disabledRestart();
	});
	var updateVibrancy = vscode.commands.registerCommand('extension.updateVibrancy', async () => {
		await Update();
		enabledRestart();
	});

	context.subscriptions.push(installVibrancy);
	context.subscriptions.push(uninstallVibrancy);
	context.subscriptions.push(updateVibrancy);

	if (isFirstload()) {
		vscode.window.showInformationMessage(localize('messages.firstload'), { title: localize('messages.installIde') })
			.then(async (msg) => {
				if (msg) {
					await Update();
					await checkColorTheme();
					enabledRestart();
				}
			});
		lockFirstload();
	}

	var lastConfig = vscode.workspace.getConfiguration("vscode_vibrancy");

	vscode.workspace.onDidChangeConfiguration(() => {
		newConfig = vscode.workspace.getConfiguration("vscode_vibrancy");
		if (!deepEqual(lastConfig, newConfig)) {
			lastConfig = newConfig;
			vscode.window.showInformationMessage(localize('messages.configupdate'), { title: localize('messages.reloadIde') })
				.then(async (msg) => {
					if (msg) {
						await Update();
						if (newConfig.theme !== vscode.workspace.getConfiguration("vscode_vibrancy")) {
							await checkColorTheme();
						}
						promptRestart();
					}
				});
			lockFirstload();
		}
	});
}
exports.activate = activate;

// this method is called when your extension is deactivated
function deactivate() { }
exports.deactivate = deactivate;
